src/mon/PGMap.cc

```diff
@@ -57,14 +59,21 @@ void PGMapDigest::encode(bufferlist& bl, uint64_t features) const
   encode(pg_pool_sum, bl, features);
   encode(pg_sum, bl, features);
   encode(osd_sum, bl, features);
-  if (v >= 2) {
+  if (v >= 5) {
     encode(num_pg_by_state, bl);
+  } else if (v >= 2) {
+    uint32_t n = num_pg_by_state.size();
+    encode(n, bl);
+    for (auto p : num_pg_by_state) {
+      encode((uint64_t)p.first, bl);
+      encode((uint32_t)p.second, bl);
+    }
   } else {
     uint32_t n = num_pg_by_state.size();
     encode(n, bl);
     for (auto p : num_pg_by_state) {
       encode((uint32_t)p.first, bl);
-      encode(p.second, bl);
+      encode((uint32_t)p.second, bl);
```

Comment:
as put in my another comment, this is not correct.

but, before a25221e55172c326b48dea6f08c16d700921864e, 
```
mempool::pgmap::unordered_map<int32_t,int32_t> num_pg_by_state;
// v1
```

after a25221e55172c326b48dea6f08c16d700921864e:

```
mempool::pgmap::unordered_map<uint64_t,int32_t> num_pg_by_state;
// v2
    uint32_t n = num_pg_by_state.size();
     ::encode(n, bl);
     for (auto p : num_pg_by_state) {
       ::encode((uint32_t)p.first, bl); // <<< buggy! should have been `int32_t`
       ::encode(p.second, bl);
     }
```

so i think we should instead encode the key (state) as int32_t to be compatible with v1 decoder. like
```c++
if (v >= 2) {
  ::encode(num_pg_by_state, bl);
} else {
  uint32_t n = num_pg_by_state.size();
  encode(n, bl);
  for (auto [state, num] : num_pg_by_state) {
    encode((int32_t)state, bl);
    encode(state, bl);
  }
}
```